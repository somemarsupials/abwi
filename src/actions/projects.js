import { api } from '../App';

export const actions = {
  FETCHING: 'FETCHING',
  FETCH_SUCCESS: 'FETCH_SUCCESS',
  FETCH_FAIL: 'FETCH_FAIL',
};

async function makeRequest() {
  return await fetch(`${api}/projects`);
};

export function projectsLoading(bool) {
  return {
    type: actions.FETCHING,
    isFetching: bool,
  };
};

export function projectsFetched(data) {
  return {
    type: actions.FETCH_SUCCESS,
    data: data,
  };
};

export function projectsNotFetched(error) {
  return {
    type: actions.FETCH_FAIL,
    error: error,
  };
};

export function fetchProjects() {
  return async (dispatch) => {
    let response, data, error;
    dispatch(projectsLoading(true));

    try {
      response = await makeRequest();
      data = await response.json();
    } 
    catch (e) {
      error = e.message;
    };

    dispatch(projectsLoading(false));

    if (!error && !response.ok) {
      error = response.ok;
    };

    if (error) {
      dispatch(projectsNotFetched(error));
    } else {
      dispatch(projectsFetched(data));
    };
  };
};
