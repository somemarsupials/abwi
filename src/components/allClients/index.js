export { default as ClientsPage } from './clientsList';
export { default as ClientsList } from './clientsPage';
