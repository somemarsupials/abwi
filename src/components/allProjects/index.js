export { default as List } from './list';
export { default as Page } from './page';
