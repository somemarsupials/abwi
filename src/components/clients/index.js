export { default as ClientPage } from './clientPage';
export { default as ClientDetail } from './clientDetail';
