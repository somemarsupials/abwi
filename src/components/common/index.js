export { default as TitleBar } from './titleBar';
export { default as ErrorBar } from './errorBar';
