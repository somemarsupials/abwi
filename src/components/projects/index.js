export { default as ProjectPage } from './projectPage';
export { default as ProjectDetail } from './projectDetail';
