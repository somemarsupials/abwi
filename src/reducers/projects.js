import { combineReducers } from 'redux';
import { actions } from '../actions/projects';
import { actions as modalActions } from '../actions/projectModal';

function fetching(state = false, action) {
  switch (action.type) {
    case actions.FETCH_LOADING:
      return action.active;
    default:
      return state;
  }
};

const initialState = {
  error: null,
  data: [],
};

function fetched(state = initialState, action) {
  switch (action.type) {
    case actions.FETCHED:
      return {
        error: action.error,
        data: action.data,
      };
    case modalActions.CREATED:
      return {
        error: state.error,
        data: state.data.concat([action.data]),
      };
    default:
      return state;
  }
};

export default combineReducers({
  fetching: fetching,
  response: fetched,
});
