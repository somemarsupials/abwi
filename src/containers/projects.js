import React, { Component } from 'react';
import { connect } from 'react-redux';

import { projectActions } from '../actions';
import { ProjectsOverview } from '../components/projects';


export class Projects extends Component {
  async componentDidMount() {
    let { api, success, fail } = this.props;
    this.props.fetch(api, success, fail);
  };

  render() {
    return (
      <ProjectsOverview
        projects={this.props.projects} 
        error={this.props.error}
      />
    );
  };
};


function mapStateToProps(state) {
  return { 
    projects: state.projects.data,
    error: state.projects.error,
  };
};


function mapDispatchToProps(dispatch) {
  let { fetchedProjects, projectFetchFailed } = projectActions;
  return {
    success: (projects) => {
      dispatch(fetchedProjects(projects))
    },
    fail: (error) => {
      dispatch(projectFetchFailed(error))
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Projects);
