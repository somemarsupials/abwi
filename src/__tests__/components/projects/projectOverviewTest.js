import React from 'react';
import { shallow } from 'enzyme';

import { ProjectsOverview } from '../../../components/projects';

describe('ProjectsOverview', () => {
  let projectsOverview;
  let projects;

  beforeEach(() => {
    projects = 'list-of-projects';
  });

  describe('without project', () => {
    beforeEach(() => {
      projectsOverview = shallow(<ProjectsOverview />);
    });

    it('renders with error bar', () => {
      expect(projectsOverview).toMatchSnapshot();
    });
  });

  describe('with error', () => {
    beforeEach(() => {
      projectsOverview = shallow(
        <ProjectsOverview error={'an error'}/>
      );
    });

    it('renders with error bar', () => {
      expect(projectsOverview).toMatchSnapshot();
    });
  });

  describe('with project and error', () => {
    beforeEach(() => {
      projectsOverview = shallow(
        <ProjectsOverview projects={projects} error={'an error'}/>
      );
    });

    it('renders with error bar', () => {
      expect(projectsOverview).toMatchSnapshot();
    });
  });

  describe('when passed project', () => {
    beforeEach(() => {
      projectsOverview = shallow(
        <ProjectsOverview projects={projects} />
      );
    });

    it('correctly renders', () => {
      expect(projectsOverview).toMatchSnapshot();
    });
  });
});

