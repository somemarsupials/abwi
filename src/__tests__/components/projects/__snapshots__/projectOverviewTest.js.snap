// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsOverview when passed project correctly renders 1`] = `
<div>
  <Component
    text="Projects"
  />
  <Component
    projects="list-of-projects"
  />
</div>
`;

exports[`ProjectsOverview with error renders with error bar 1`] = `
<Component
  detail="an error"
  text="Could not find projects"
/>
`;

exports[`ProjectsOverview with project and error renders with error bar 1`] = `
<Component
  detail="an error"
  text="Could not find projects"
/>
`;

exports[`ProjectsOverview without project renders with error bar 1`] = `
<Component
  text="Could not find projects"
/>
`;
